File: CPlusPlus/Problem 006/Problem006.cpp
/*
Project Euler
Problem 6
"Find the difference between the sum of the squares of the first one hundred natural numbers and the square of the sum."
*/

#include <iostream>

int main (void)
{
	// the sum of the numbers 1 through n is (n * (n + 1))/2.  
	// so the square of that sum is (n^2 * (n + 1) ^2)/4
	long square_sum = 100 * 100 * 101 * 101 / 4;

	// calculate the sum of squares iteratively
	long sum_squares = 0;
	for (int i=1; i<=100; ++i)
	{
		sum_squares += i * i;
	}

	std::cout << sum_squares - square_sum << std::endl;
}

File: Scala/src/main/scala/euler/Problem006.scala
package euler

// Project Euler
// Problem 6
//
// The sum of the squares of the first ten natural numbers is,
//
// 1^2 + 2^2 + ... + 10^2 = 385
// The square of the sum of the first ten natural numbers is,
//
// (1 + 2 + ... + 10)^2 = 552 = 3025
// Hence the difference between the sum of the squares of the first ten natural numbers and the
// square of the sum is 3025  385 = 2640.
//
// Find the difference between the sum of the squares of the first one hundred natural numbers and
// the square of the sum.

object Problem006 {
	def main (args: Array[String]) = {
		val nums = List.range(1L, 101L);
		val squares = nums.map(x => x*x);
		val sum = nums.sum;
		val sum_of_squares = squares.sum;
		val square_of_sum = sum*sum;
		println(square_of_sum - sum_of_squares);
	}
}